File: plugins/rating/glicko2.lua
-- plugins/rating/glicko2.lua
-- Full Glicko-2 (Glickman 2012): scale to (mu, phi, sigma), per-opponent g/E,
-- v and Delta, Newton-Raphson volatility iteration, then scale back.
-- config: initial_rating, initial_rd, initial_volatility, tau, epsilon
-- Verified against the paper worked example (r'=1464.06, RD'=151.52, sigma'=0.05999).

information_budget = { "WinLoss" }

local SCALE = 173.7178
local RATING_CENTER = 1500.0

function initialize(player_id, config, context)
    return {
        rating = config.initial_rating,
        rating_deviation = config.initial_rd,
        volatility = config.initial_volatility,
        games_played = 0,
    }, context
end

function predict(team_a, team_b, config, context)
    local avg_a = team_average(team_a)
    local avg_b = team_average(team_b)
    return 1.0 / (1.0 + math.exp(-(avg_a - avg_b) / 400.0))
end

function team_average(team)
    local sum = 0.0
    for _, o in ipairs(team) do
        sum = sum + o.rating
    end
    if #team == 0 then return 0.0 end
    return sum / #team
end

function g(phi)
    return 1.0 / math.sqrt(1.0 + 3.0 * phi * phi / (math.pi * math.pi))
end

function e(mu, mu_j, phi_j)
    local gj = g(phi_j)
    return 1.0 / (1.0 + math.exp(-gj * (mu - mu_j)))
end

-- Newton-Raphson on Glicko-2's f(x) to find the new volatility (steps 5.2-5.6).
function new_volatility(sigma, delta, phi, v, tau, epsilon)
    local a = math.log(sigma * sigma)
    local function big_f(x)
        local ex = math.exp(x)
        return (ex * (delta * delta - phi * phi - v - ex))
                / (2.0 * (phi * phi + v + ex) * (phi * phi + v + ex))
            - (x - a) / (tau * tau)
    end

    local b
    if delta * delta > phi * phi + v then
        b = math.log(delta * delta - phi * phi - v)
    else
        local k = 1
        while big_f(a - k * tau) < 0.0 do
            k = k + 1
        end
        b = a - k * tau
    end

    local fa = big_f(a)
    local fb = big_f(b)
    local a_val = a
    local b_val = b
    while math.abs(b_val - a_val) > epsilon do
        local c = a_val + (a_val - b_val) * fa / (fb - fa)
        local fc = big_f(c)
        if fc * fb <= 0.0 then
            a_val = b_val
            fa = fb
        else
            fa = fa / 2.0
        end
        b_val = c
        fb = fc
    end

    -- Converged x = ln(sigma^2); sigma' = exp(x/2) = exp((a+b)/4).
    return math.exp((a_val + b_val) / 4.0)
end

function scale(rating, rd)
    return (rating - RATING_CENTER) / SCALE, rd / SCALE
end

function unscale(mu, phi)
    return RATING_CENTER + SCALE * mu, SCALE * phi
end

function update_player(mu, phi, sigma, opponents, epsilon, tau)
    local v_inv = 0.0
    local delta_numer = 0.0
    for _, opp in ipairs(opponents) do
        local mu_j, phi_j, outcome = opp[1], opp[2], opp[3]
        local gj = g(phi_j)
        local e_val = e(mu, mu_j, phi_j)
        v_inv = v_inv + gj * gj * e_val * (1.0 - e_val)
        delta_numer = delta_numer + gj * (outcome - e_val)
    end
    if v_inv == 0.0 then
        return mu, phi, sigma
    end
    local v = 1.0 / v_inv
    local delta = v * delta_numer
    local sigma_prime = new_volatility(sigma, delta, phi, v, tau, epsilon)
    local phi_star = math.sqrt(phi * phi + sigma_prime * sigma_prime)
    local phi_prime = 1.0 / math.sqrt(1.0 / (phi_star * phi_star) + 1.0 / v)
    local mu_prime = mu + phi_prime * phi_prime * delta_numer
    return mu_prime, phi_prime, sigma_prime
end

function update(match_result, observations, config, context)
    local epsilon = config.epsilon or 0.000001
    local tau = config.tau or 0.5
    local team_a = match_result.team_a
    local team_b = match_result.team_b
    local team_a_won = match_result.winner == "A"
    local outcome_a = team_a_won and 1.0 or 0.0
    local outcome_b = 1.0 - outcome_a

    local function collect_opponents(ids, outcome)
        local opponents = {}
        for _, id in ipairs(ids) do
            local o = observations[id]
            if o then
                local mu, phi = scale(o.rating, o.rating_deviation)
                table.insert(opponents, { mu, phi, outcome })
            end
        end
        return opponents
    end

    local opp_b = collect_opponents(team_b, outcome_a)
    local opp_a = collect_opponents(team_a, outcome_b)

    local updates = {}
    local function update_team(ids, opponents)
        for _, id in ipairs(ids) do
            local o = observations[id]
            if o then
                local mu, phi = scale(o.rating, o.rating_deviation)
                local mu_p, phi_p, sigma_p = update_player(mu, phi, o.volatility,
                                                           opponents, epsilon, tau)
                local rating, rd = unscale(mu_p, phi_p)
                if config.floor and config.ceiling then
                    rating = math.max(config.floor, math.min(config.ceiling, rating))
                end
                table.insert(updates, {
                    player_id = id,
                    rating = rating,
                    rating_deviation = rd,
                    volatility = sigma_p,
                    games_played = o.games_played + 1,
                })
            end
        end
    end

    update_team(team_a, opp_b)
    update_team(team_b, opp_a)
    return updates, context
end

File: crates/matchlab-validation/src/reference/glicko2.rs
//! Independent Glicko-2 reference implementation.
//!
//! Derived from Mark Glickman, "Example of the Glicko-2 system" (2012):
//! <https://www.glicko.net/glicko/glicko2.pdf>. Equations are cited inline
//! (eq. 1-10). This exists only to catch drift in `plugins/rating/glicko2.lua`
//! from *outside* the Lua stack — when they disagree, the script is the bug.

/// The scale constant (eq. 1-2): ratings live on a scale where 173.7178 units
/// ≈ 1 unit of "deviation-scale" skill.
pub const SCALE: f64 = 173.7178;
/// The rating center the paper shifts to μ = 0.
pub const RATING_CENTER: f64 = 1500.0;

/// g(φ) (eq. 3): shrinks an opponent's deviation contribution.
pub fn g(phi: f64) -> f64 {
    1.0 / (1.0 + 3.0 * phi * phi / std::f64::consts::PI.powi(2)).sqrt()
}

/// E[μ, μ_j, φ_j] (eq. 4): expected score against opponent j.
pub fn e(mu: f64, mu_j: f64, phi_j: f64) -> f64 {
    1.0 / (1.0 + (-g(phi_j) * (mu - mu_j)).exp())
}

/// A single rated game in a rating period.
#[derive(Debug, Clone, Copy)]
pub struct Opponent {
    pub mu: f64,
    pub phi: f64,
    /// 1.0 for a win, 0.0 for a loss.
    pub outcome: f64,
}

/// Result of one rating period on the Glicko scale.
#[derive(Debug, Clone, Copy)]
pub struct PeriodResult {
    pub mu: f64,
    pub phi: f64,
    pub sigma: f64,
}

/// The f(x) helper of the volatility iteration (eq. 7).
fn big_f(x: f64, delta: f64, phi: f64, v: f64, a: f64, tau: f64) -> f64 {
    let ex = x.exp();
    (ex * (delta * delta - phi * phi - v - ex))
        / (2.0 * (phi * phi + v + ex) * (phi * phi + v + ex))
        - (x - a) / (tau * tau)
}

/// One full Glicko-2 rating period for a single player against a list of
/// opponents with the same outcome semantics as Glickman's paper (steps 2-4
/// plus 5.1: the players' ratings entering the period use the *pre-period*
/// σ in φ*; the volatility iteration is Newton's method with the
/// bisection-style safeguard of eq. 7-9).
pub fn single_period(
    mu: f64,
    phi: f64,
    sigma: f64,
    opponents: &[Opponent],
    tau: f64,
    epsilon: f64,
) -> PeriodResult {
    let v_inv: f64 = opponents
        .iter()
        .map(|o| {
            let gj = g(o.phi);
            let e_j = e(mu, o.mu, o.phi);
            gj * gj * e_j * (1.0 - e_j)
        })
        .sum::<f64>();
    if v_inv == 0.0 {
        return PeriodResult { mu, phi, sigma };
    }
    let v = 1.0 / v_inv;
    let delta: f64 = v * opponents
        .iter()
        .map(|o| g(o.phi) * (o.outcome - e(mu, o.mu, o.phi)))
        .sum::<f64>();

    // Step 5.2-5.6: solve f(x) = 0 for x = ln σ².
    let a = (sigma * sigma).ln();
    let b_start = if delta * delta > phi * phi + v {
        (delta * delta - phi * phi - v).ln()
    } else {
        let mut k = 1.0;
        while big_f(a - k * tau, delta, phi, v, a, tau) < 0.0 {
            k += 1.0;
        }
        a - k * tau
    };

    let mut fa = big_f(a, delta, phi, v, a, tau);
    let mut fb = big_f(b_start, delta, phi, v, a, tau);
    let (mut a_val, mut b_val) = (a, b_start);
    while (b_val - a_val).abs() > epsilon {
        let c = a_val + (a_val - b_val) * fa / (fb - fa);
        let fc = big_f(c, delta, phi, v, a, tau);
        if fc * fb <= 0.0 {
            a_val = b_val;
            fa = fb;
        } else {
            fa /= 2.0;
        }
        b_val = c;
        fb = fc;
    }
    let sigma_prime = ((a_val + b_val) / 4.0).exp();

    // Step 6-7: φ* from φ² + σ'²; φ' from combining φ* and v; μ' from Δ.
    let phi_star = (phi * phi + sigma_prime * sigma_prime).sqrt();
    let phi_prime = 1.0 / (1.0 / (phi_star * phi_star) + 1.0 / v).sqrt();
    let delta_numer: f64 = opponents
        .iter()
        .map(|o| g(o.phi) * (o.outcome - e(mu, o.mu, o.phi)))
        .sum::<f64>();
    let mu_prime = mu + phi_prime * phi_prime * delta_numer;

    PeriodResult {
        mu: mu_prime,
        phi: phi_prime,
        sigma: sigma_prime,
    }
}

/// Scale a rating-period input to the Glicko scale (eq. 1-2).
pub fn scale(rating: f64, rd: f64) -> (f64, f64) {
    ((rating - RATING_CENTER) / SCALE, rd / SCALE)
}

/// Scale a Glicko-scale result back to rating units (eq. 10).
pub fn unscale(mu: f64, phi: f64) -> (f64, f64) {
    (RATING_CENTER + SCALE * mu, SCALE * phi)
}

/// Volatility growth over an idle period (the φ* step with no games):
/// φ' = sqrt(φ² + σ²), rating unchanged. Glicko-2 applies this between rating
/// periods for players who did not play; the Lua script has no idle entry
/// point, so this reference is used to compose the T-02 two-period chain
/// (the grown RD is fed back into the script as ordinary pre-period input).
pub fn idle_step(phi: f64, sigma: f64) -> f64 {
    (phi * phi + sigma * sigma).sqrt()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn paper_worked_example_single_period() {
        // Player 1500/200/0.06 (scaled μ=0, φ=200/173.7178, σ=0.06) plays one
        // rating period: win vs 1400/30, loss vs 1550/100, loss vs 1700/300.
        let (mu, phi) = scale(1500.0, 200.0);
        let (m1400, p1400) = scale(1400.0, 30.0);
        let (m1550, p1550) = scale(1550.0, 100.0);
        let (m1700, p1700) = scale(1700.0, 300.0);
        let opponents = [
            Opponent {
                mu: m1400,
                phi: p1400,
                outcome: 1.0,
            },
            Opponent {
                mu: m1550,
                phi: p1550,
                outcome: 0.0,
            },
            Opponent {
                mu: m1700,
                phi: p1700,
                outcome: 0.0,
            },
        ];
        let out = single_period(mu, phi, 0.06, &opponents, 0.5, 1e-6);
        let (r, rd) = unscale(out.mu, out.phi);
        assert!((r - 1464.06).abs() < 0.01, "r {r}");
        assert!((rd - 151.52).abs() < 0.01, "rd {rd}");
        assert!((out.sigma - 0.05999).abs() < 1e-5, "sigma {}", out.sigma);
    }
}
#[cfg(test)]
mod probe_tests {
    use super::*;
    #[test]
    #[ignore]
    fn probe_single_win() {
        let (mu, phi) = scale(1500.0, 200.0);
        let (mu_j, phi_j) = scale(1400.0, 30.0);
        let out = single_period(
            mu,
            phi,
            0.06,
            &[Opponent {
                mu: mu_j,
                phi: phi_j,
                outcome: 1.0,
            }],
            0.5,
            1e-6,
        );
        let (r, rd) = unscale(out.mu, out.phi);
        println!("SINGLE-WIN REF => r={r} rd={rd} sigma={}", out.sigma);
    }
}
